### PR TITLE
FEATURE: 1926 cea worker / non-blocking tools

### PR DESCRIPTION
**Pull request by daren-thomas**
This PR addresses #1926 (As a user I would like to not get blocked once one Job in the CEA dashboard starts) and more:

- When you start a tool, instead of showing the tool output in a modal dialog, you're taken to an overview of "jobs" run in this session. 
- Each such job has a status (pending, started, success/error), a script name (what was run) and the parameters used for the job. They are listed in a table. 
- Clicking on a row will open up the dialog with the job output. This works for jobs that are running as well as those that have completed. 
- You can start many jobs, the outputs are collected independently
- While jobs are running, a ticker at the bottom right of the page shows the last line output - this updates as jobs run and also includes the job id of the line being shown.

In addition, the creation of the `cea-worker` process is one step closer to working on the cloud - communication happens over socket.io, so in theory, this could also work in a distributed fashion. There is still the issue of accessing data, though, that hasn't been resolved yet.

To run this, you'll need to do a `pip install -e .` in the source folder as the `cea-worker` script is a pip entry point (like `cea`, `cea-config` and `cea-doc`)

@JIMENOFONSECA is this what you had in mind?

@reyery the output ticker at the bottom is just the simplest solution I could think of - but can easily be wired into any other container, say, in a hypothetical SPA.

### CODE REVIEW

**Comment by reyery** [comment]
I was able to run the tools and see the tools table but was not able to get `socketio` functions to work i.e. the ticker at the bottom. For some reason I am getting an error when it tries to run `socketio.emit()`, as `socketio.server` seems to be `None`. It's as if  `socketio` has not initialized with the Flask `app`.

```
[2019-09-09 13:40:37,720] ERROR in app: Exception on /server/jobs/started/1 [POST]
Traceback (most recent call last):
  File "C:\Users\Admin\Documents\CityEnergyAnalyst\Dependencies\Python\lib\site-packages\flask\app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "C:\Users\Admin\Documents\CityEnergyAnalyst\Dependencies\Python\lib\site-packages\flask\app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "C:\Users\Admin\Documents\CityEnergyAnalyst\Dependencies\Python\lib\site-packages\flask_restplus\api.py", line 325, in wrapper
    resp = resource(*args, **kwargs)
  File "C:\Users\Admin\Documents\CityEnergyAnalyst\Dependencies\Python\lib\site-packages\flask\views.py", line 88, in view
    return self.dispatch_request(*args, **kwargs)
  File "C:\Users\Admin\Documents\CityEnergyAnalyst\Dependencies\Python\lib\site-packages\flask_restplus\resource.py", line 44, in dispatch_request
    resp = meth(*args, **kwargs)
  File "C:\Users\Admin\Documents\CityEnergyAnalyst\Dependencies\Python\lib\site-packages\flask_restplus\marshalling.py", line 136, in wrapper
    resp = f(*args, **kwargs)
  File "C:\Users\Admin\Documents\CityEnergyAnalyst\CityEnergyAnalyst\cea\interfaces\dashboard\server\jobs.py", line 105, in post
    socketio.emit("cea-worker-started", api.marshal(job, job_info_model))
AttributeError: 'NoneType' object has no attribute 'emit'
```

https://github.com/architecture-building-systems/CityEnergyAnalyst/blob/5e1e480a78f7de23e82283dfcb9826bc2b93b1d3/cea/interfaces/dashboard/server/jobs.py#L99-L106

**Comment by daren-thomas** [comment]
@reyery how are you running the dashboard? I've tried both methods: Running from CEA Console with `cea dashboard` and launching the CEA Dashboard shortcut. Both work, showing the ticker.

The behavior you're describing suggests you also don't see the script output? (because that works with the same socketio stuff?)

There are two relevant parts in `dashboard.main`:
- `global socketio, socketio = SocketIO(app)`
- `socketio.run(app, host='localhost', port=5050)`

Are these present in the version you're running?

**Comment by reyery** [comment]
@daren-thomas ah running it from Console or shortcut works for me, was running it from PyCharm previously and it still gives that error now.

**Comment by daren-thomas** [comment]
So it works?

(it doesn't work from PyCharm for me either - but I think that's due to issues with the webserver used... it needs to be the `gevent` server. maybe flask defaults to the flask server when running in PyCharm and that doesn't handle the threads properly for socketio to work)

BTW: running from PyCharm does not need to be supported anyway.

**Comment by JIMENOFONSECA** [comment]
It does matter right? part of the developing team (engineers) use Pycharm as the main interface to build the back-end and check the outputs in the dashboard. Blocking the use of Pycharm makes their job difficult, do not you think?

On 10 Sep 2019, 7:08 PM +0800, Daren Thomas <notifications@github.com>, wrote:

So it works?

(it doesn't work from PyCharm for me either - but I think that's due to issues with the webserver used... it needs to be the gevent server. maybe flask defaults to the flask server when running in PyCharm and that doesn't handle the threads properly for socketio to work)

BTW: running from PyCharm does not need to be supported anyway.

**Comment by daren-thomas** [comment]
@JIMENOFONSECA well - neither background development nor dashboard checking would have been impacted - just viewing output from scripts in the ticker... but I did find a way to have it work in PyCharm too :)

@reyery would you mind testing again? once this is merged, we can think about improving the look and feel of the interface (i view this PR as a "plumbing" thing - we still need to add bells and whistles but that kind of depends on the reacted interface)

File: cea/interfaces/dashboard/base/static/cea-worker-ticker.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    // this is a callback that triggers when the "my response" event is emitted by the server.
+    socket.on("cea-worker-message", function (data) {
+        let lines = data.message.split(/\r?\n/).map(x => x.trim()).filter(x => x.length > 0);
+        let last_line = lines[lines.length - 1];
+
+        $ticker.text(`${data.jobid}: ${last_line.substr(0, 80)}`);
+    });
+
+    socket.on("cea-worker-success", function (data) {
+        $ticker.text(`${data.jobid}: completed`);
```
[the remark this answers is not in the document]
Comment (daren-thomas) [reply cea/interfaces/dashboard/base/static/cea-worker-ticker.js]: you're right. i fixed it. thank you. ready to merge, @reyery?

**Pull request merged**